### DISCUSSION TITLE
Build Output and Dockerfile COPY

### DISCUSSION THREAD

**Discussion by dschinkel** [community/community#26034] [Actions]
<p>My build outputs a <strong>dist</strong> directory via <strong>webpack</strong>.</p>
<p><strong>Problem:</strong> I’ve tried some COPY commands from the <strong>dist</strong> directory of my <strong>build</strong> but for some reason I’m not doing it right and the copy to my Docker container fails.</p>
<p><strong>I get:</strong></p>
<pre><code class="lang-auto">Step 3/3 : COPY ./dist ./dist

(https://github.com/xxxx/xxxx/runs/1048806314?check_suite_focus=true#step:4:47)COPY failed: stat /var/lib/docker/tmp/docker-builder088552185/dist: no such file or directory`

</code></pre>
<p>Example <strong>Dockerfile</strong> where I cannot get copy to work:</p>
<pre><code class="lang-auto">FROM node:14.8.0-alpine
ENV PORT 80
COPY ./dist ./dist
</code></pre>
<p>also tried:</p>
<pre><code class="lang-auto">COPY dist dist
</code></pre>
<p>and</p>
<pre><code class="lang-auto">COPY ./dist dist
</code></pre>
<p>and</p>
<pre><code class="lang-auto">COPY dist ./dist
</code></pre>
<p>I know that the files are there because I ran an <code>ls -la</code> step after my build was done to list out the output of the <code>$GITHUB_WORKSPACE</code>.<br>
<img src="https://user-images.githubusercontent.com/338798/181094891-46b65799-6ad4-431b-ba38-8d4253fc54e7.jpeg" alt="Screen Shot 2020-08-30 at 6.12.33 PM" data-base62-sha1="vkVvdDGM62AFlYZ0bTsATmtRAls" width="150" height="150"></p>
<p>I already read <a href="https://docs.github.com/en/actions/creating-actions/dockerfile-support-for-github-actions" rel="nofollow noopener">Dockerfile support for GitHub Actions</a> but I must be missing something obvious, just need another set of eyes</p>
<p><strong>Build.yml</strong></p>
<pre><code class="lang-auto">name: build-test

on:
  push:
    branches:
      - master
  pull_request:
    branches:
      - master
jobs:
  build:
    name: Build WeDoTDD Server &amp; Client
    runs-on: ubuntu-latest
    steps:
      - uses: actions/checkout@v2
      - name: Setup node 14
        uses: actions/setup-node@v1
        with:
          node-version: 14.x
      - run: yarn --frozen-lockfile
      - run: yarn build-server
      - run: yarn build-client
      - run: ls ./dist
        env:
          CI: true
  push_to_registry:
    needs: build
    name: Push Docker image to GitHub Packages
    runs-on: ubuntu-latest
    steps:
      - name: Check out the repo
        uses: actions/checkout@v2
      - name: Push to GitHub Packages
        uses: docker/build-push-action@v1
        with:
          username: ${{ github.actor }}
          password: ${{ secrets.GITHUB_TOKEN }}
          registry: docker.pkg.github.com
          # OWNER/REPOSITORY/IMAGE_NAME
          repository: xxxx/xxxx/staging
          tag_with_sha: true
</code></pre>

**Comment by thinkverse** [comment]
<p>Hi. :wave:</p>
<p>My knowledge of Docker is limited, but looking at the documentation for <a href="https://docs.docker.com/engine/reference/builder/#copy" rel="nofollow noopener">COPY</a>, it copies files from the <code>&lt;src&gt;</code> into the container at the given <code>&lt;dest&gt;</code>, now I’m not 100% sure if the output from the <code>run</code> command in a GitHub Actions, places the build in the <code>&lt;src&gt;</code> that Docker uses, but if it does, then I’m guessing the COPY you want would be this.</p>
<pre><code class="lang-auto">COPY dist /dist
</code></pre>
<p>Which from my understanding would copy the contents of the <code>dist</code> source folder to the <code>/dist</code> folder in the container. :thinking:</p>
<p>Hope that helps. :+1:</p>

**Comment by dschinkel** [comment]
<p>Thanks tried it but no luck.  I wonder if it’s because I’m calling <code>actions/checkout@v2</code> twice and the second time wipes out the workflow?</p>
<p>Yea the docs on <a href="https://github.com/docker/build-push-action" rel="nofollow noopener">build-push-action</a> are lacking as to what the context is that the copy is done <em>from</em>.</p>

**Comment by thinkverse** [comment, marked as answer]
<p>I think the issue is that you’re using two jobs, so the build from one job isn’t shared to the next job, what you wanna use <a href="https://github.com/actions/upload-artifact" rel="nofollow noopener">actions/upload-artifact</a> and <a href="https://github.com/actions/download-artifact" rel="nofollow noopener">actions/download-artifact</a> to share the build between jobs, there was also some changes to the <a href="https://github.blog/changelog/2020-01-14-github-actions-changes-to-artifact-download-experience/" rel="nofollow noopener">download experience</a> you would wanna read as well.</p>

**Comment by dschinkel** [comment]
<p>Thanks I’ll check that out.</p>
<p>I assume I’d use download instead of <code>checkout</code> on the second job.</p>

**Comment by dschinkel** [comment]
<p>Got it working.  First part was getting the artifact transfered, so thanks much!</p>
<p>Second part was getting the Dockerfile right:</p>
<p><strong>Dockerfile</strong> - I added WORKDIR and then used current directory off the working directory:</p>
<pre><code class="lang-auto">FROM node:14.8.0-alpine
ENV PORT 80
WORKDIR dist
COPY . dist
</code></pre>
<p>Thanks for the tip on the artifacts!</p>

**Comment by thinkverse** [comment]
<p>Congratulations. :tada:</p>
<p>I’m glad I could provide some help that led to your solution, I wish you luck in further endeavors. :slightly_smiling_face:</p>
